### PR TITLE
K8SPG-345: Allowing exposing superusers through pgbouncer

### PR DESCRIPTION
**Pull request by egegunes**
[K8SPG-345](https://jira.percona.com/browse/K8SPG-345) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

**CHANGE DESCRIPTION**
---
**Problem:**
By default `auth_query` in PGBouncer doesn't allow SUPERUSERs to connect. It's the correct behavior for security but we want to give our users an option to change it.

**Solution:**
We added `exposeSuperusers` to both our and Crunchy's PGBouncer spec in CR. If the value is true, we remove the condition from `auth_query` that blocks SUPERUSERs to connect.

**CHECKLIST**
---
**Jira**
- [x] Is the Jira ticket created and referenced properly?
- [x] Does the Jira ticket have the proper statuses for documentation (`Needs Doc`) and QA (`Needs QA`)?
- [x] Does the Jira ticket link to the proper milestone (Fix Version field)?

**Tests**
- [x] Is an E2E test/test case added for the new feature/change?
- [x] Are unit tests added where appropriate?

**Config/Logging/Testability**
- [x] Are all needed new/changed options added to default YAML files?
- [x] Are the manifests (crd/bundle) regenerated if needed?
- [x] Did we add proper logging messages for operator actions?
- [x] Did we ensure compatibility with the previous version or cluster upgrade process?
- [x] Does the change support oldest and newest supported PG version?
- [x] Does the change support oldest and newest supported Kubernetes version?

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/demand-backup.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/scaling.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/scheduled-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/monitoring.log) |
| telemetry-transfer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/telemetry-transfer.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/2b2ed144a/users.log) |
| We run 8 out of 8|

commit: https://github.com/percona/percona-postgresql-operator/pull/465/commits/2b2ed144af1f9efe7f4e43892d2512d25ff576c9
image: `perconalab/percona-postgresql-operator:PR-465-2b2ed144a`

File: deploy/cr.yaml [hunk 1]
```diff
@@ -20,7 +20,9 @@ spec:
 
 #  openshift: true
 
-#  users:
+  users:
```
Comment (hors) [inline deploy/cr.yaml:23]: please remove it

File: e2e-tests/build [hunk 2]
```diff
@@ -35,7 +35,6 @@ build_operator() {
 		--build-arg GO_LDFLAGS="${GO_LDFLAGS}" \
 		$squash \
 		$no_cache \
-		--progress plain \
```
Comment (hors) [inline e2e-tests/build:38]: Why was it removed?

**Review by hors — CHANGES_REQUESTED** [review]

**Review by inelpandzic — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/demand-backup.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/scaling.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/scheduled-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/monitoring.log) |
| telemetry-transfer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/telemetry-transfer.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/f3fac4859/users.log) |
| We run 8 out of 8|

commit: https://github.com/percona/percona-postgresql-operator/pull/465/commits/f3fac4859a9e8b47e762feacdb420eb1797d1c65
image: `perconalab/percona-postgresql-operator:PR-465-f3fac4859`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/demand-backup.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/scaling.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/scheduled-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/monitoring.log) |
| telemetry-transfer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/telemetry-transfer.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/ec1cf2407/users.log) |
| We run 8 out of 8|

commit: https://github.com/percona/percona-postgresql-operator/pull/465/commits/ec1cf2407d87cde0b60e0b24b21a98a37c58818f
image: `perconalab/percona-postgresql-operator:PR-465-ec1cf2407`

File: internal/pgbouncer/postgres.go [hunk 3]
```diff
@@ -55,7 +55,14 @@ func sqlAuthenticationQuery(sqlFunctionName string) string {
...
 		`(pg_authid.rolvaliduntil IS NULL OR pg_authid.rolvaliduntil >= CURRENT_TIMESTAMP)`,
-	}, "\n    AND ")
+	}
+
+	// Remove NOT pg_authid.rolsuper
+	if exposeSuperusers {
+		conditions = append(conditions[:1], conditions[2:]...)
+	}
+
+	sqlAuthorizationConditions := strings.Join(conditions, "\n    AND ")
```
Comment (hors) [inline internal/pgbouncer/postgres.go:65]: but the main idea is to allow postgres user to connect to the DB via bgbouncer but it does not work now

**Review by hors — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/demand-backup.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/scaling.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/scheduled-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/monitoring.log) |
| telemetry-transfer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/telemetry-transfer.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/9f7d28548/users.log) |
| We run 8 out of 8|

commit: https://github.com/percona/percona-postgresql-operator/pull/465/commits/9f7d285482f4b81b8b71219eee46c8ceca51ce76
image: `perconalab/percona-postgresql-operator:PR-465-9f7d28548`

**Review by hors — APPROVED** [review]

File: internal/pgbouncer/postgres.go (hunk 3 above)
Comment (hors) [reply internal/pgbouncer/postgres.go:68]: fixed

**Review by hors — APPROVED** [review]

**Review by inelpandzic — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/init-deploy.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/demand-backup.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/scaling.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/scheduled-backup.log) |
| start-from-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/start-from-backup.log) |
| monitoring | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/monitoring.log) |
| telemetry-transfer | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/telemetry-transfer.log) |
| users | [failed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-pg-operator/PR-465/20ef02bf2/users.log) |
| We run 8 out of 8|

commit: https://github.com/percona/percona-postgresql-operator/pull/465/commits/20ef02bf236b8e78dd29dc08c5c08e688574024f
image: `perconalab/percona-postgresql-operator:PR-465-20ef02bf2`

**Pull request merged**